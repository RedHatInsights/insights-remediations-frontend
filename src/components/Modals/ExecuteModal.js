/* eslint-disable camelcase */
import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import {
  Button,
  Modal,
  ModalVariant,
  TextContent,
  Text,
  TextVariants,
  Alert,
  Tooltip,
  ExpandableSection,
  List,
  ListItem,
} from '@patternfly/react-core';
import { downloadPlaybook } from '../../api';
import { styledConnectionStatus } from '../statusHelper';
import {
  TableHeader,
  Table,
  TableBody,
  TableVariant,
} from '@patternfly/react-table';
import { Skeleton } from '@redhat-cloud-services/frontend-components/Skeleton';
import './ExecuteModal.scss';

export const ExecuteModal = ({
  isOpen,
  onClose,
  showRefresh,
  isLoading,
  data,
  remediationId,
  issueCount,
  runRemediation,
  etag,
  getEndpoint,
  sources,
  setEtag,
}) => {
  const [isUserEntitled, setIsUserEntitled] = useState(false);
  const [connected, setConnected] = useState([]);
  const [disconnected, setDisconnected] = useState([]);
  const isDebug = () => localStorage.getItem('remediations:debug') === 'true';

  useEffect(() => {
    window.insights.chrome.auth
      .getUser()
      .then((user) =>
        setIsUserEntitled(user.entitlements.smart_management.is_entitled)
      );
  }, []);

  const combineStatuses = (status, availability) =>
    status === 'connected'
      ? availability
        ? availability.availability_status
        : 'loading'
      : status;

  useEffect(() => {
    const [con, dis] = data.reduce(
      ([pass, fail], e) =>
        e && e.connection_status === 'connected'
          ? [[...pass, { ...e }], fail]
          : [pass, [...fail, e]],
      [[], []]
    );
    setConnected(con);
    setDisconnected(dis);
    con.map((c) => c.endpoint_id && getEndpoint(c.endpoint_id));
  }, [data]);

  useEffect(() => {
    const isAvailable = (connectionStatus, sourcesStatus, data) =>
      combineStatuses(
        connectionStatus,
        sourcesStatus === 'fulfilled' && data
      ) === 'available';

    const updatedData = data.map((e) => ({
      ...e,
      connection_status: combineStatuses(
        e.connection_status,
        sources.status === 'fulfilled' && sources.data[`${e.endpoint_id}`]
      ),
    }));

    if (sources.status === 'fulfilled') {
      const [con, dis] = updatedData.reduce(
        ([pass, fail], e) =>
          isAvailable(
            e.connection_status,
            sources.status,
            sources.data[`${e.endpoint_id}`]
          )
            ? [[...pass, { ...e }], fail]
            : [pass, [...fail, { ...e }]],
        [[], []]
      );
      setConnected(con);
      setDisconnected(dis);
    }
  }, [sources]);

  const generateRowsStatus = (con) => {
    return styledConnectionStatus(
      con.connection_status,
      sources.status === 'fulfilled' &&
        sources.data[`${con.endpoint_id}`] &&
        sources.data[`${con.endpoint_id}`].availability_status_error
    );
  };

  const rows = [...connected, ...disconnected].map((con) => ({
    cells: [
      {
        title: con.executor_name ? (
          <Tooltip content={`${con.executor_name}`}>
            <span>
              {con.executor_name.length > 25
                ? `${con.executor_name.slice(0, 22)}...`
                : con.executor_name}
            </span>
          </Tooltip>
        ) : (
          'Direct connection'
        ),
      },
      con.system_count,
      isUserEntitled && {
        title: generateRowsStatus(con),
      },
    ],
  }));
  const connectedCount = connected.reduce((acc, e) => e.system_count + acc, 0);
  const systemCount = data.reduce((acc, e) => e.system_count + acc, 0);

  const pluralize = (number, str) =>
    number > 1 ? `${number} ${str}s` : `${number} ${str}`;

  return (
    <Modal
      className="remediations ins-c-execute-modal"
      variant={isDebug() ? ModalVariant.large : ModalVariant.small}
      title={'Execute playbook'}
      isOpen={isOpen}
      onClose={onClose}
      isFooterLeftAligned
      actions={[
        <Button
          key="confirm"
          variant="primary"
          isDisabled={connected.length === 0}
          onClick={() => {
            runRemediation(
              remediationId,
              etag,
              disconnected.map((e) => e.executor_id).filter((e) => e)
            );
          }}
        >
          {isLoading
            ? 'Execute playbook'
            : `Execute playbook on ${pluralize(connectedCount, 'system')}`}
        </Button>,
        <Button
          key="download"
          variant="secondary"
          onClick={() => downloadPlaybook(remediationId)}
        >
          Download playbook
        </Button>,
        isDebug() ? (
          <Button key="reset-etag" onClick={() => setEtag('test')}>
            Reset etag
          </Button>
        ) : null,
      ]}
    >
      <div>
        {showRefresh ? (
          <Alert
            variant="warning"
            isInline
            title="The connection status of systems associated with this Playbook has changed. Please review again."
          />
        ) : null}
        <TextContent>
          {isLoading ? (
            <Skeleton size="lg" />
          ) : (
            <Text component={TextVariants.p}>
              Playbook contains <b>{`${pluralize(issueCount, 'action')}`}</b>
              &nbsp;affecting
              <b> {`${pluralize(systemCount, 'system')}.`} </b>
            </Text>
          )}
          <Text>
            <ExpandableSection toggleText="About remote execution with Cloud connector">
              Playbooks can be executed on systems which:
              <List>
                <ListItem>
                  Are connected to Insights via a Satellite instance which has
                  Receptor/Cloud Connector enabled, or <br />
                  <Button
                    className="pf-u-p-0"
                    key="download"
                    variant="link"
                    component="a"
                    // eslint-disable-next-line max-len
                    href="https://access.redhat.com/documentation/en-us/red_hat_insights/2020-04/html/remediating_issues_across_your_red_hat_satellite_infrastructure_using_red_hat_insights/configuring-your-satellite-infrastructure-to-communicate-with-insights"
                  >
                    How to configure Receptor/Cloud Connector on Red Hat
                    Satellite
                  </Button>
                </ListItem>
                <ListItem>
                  Are directly connected to Insights via Red Hat connector, and
                  Cloud Connector is enabled <br />
                  <Button
                    className="pf-u-p-0"
                    key="configure"
                    variant="link"
                    // eslint-disable-next-line max-len
                    href="#"
                  >
                    How to enable Cloud Connector with Red Hat connect
                  </Button>
                </ListItem>
              </List>
            </ExpandableSection>
          </Text>
          <Text component={TextVariants.p}>
            Executed Ansible Playbooks run on eligible systems with Cloud
            Connector. The playbook will be pushed immediately after selecting
            “Execute playbook”. If the playbook has “Auto reboot” on, systems
            requiring reboot to complete an action will reboot.
          </Text>
          <Button
            className="pf-u-p-0"
            key="configure"
            variant="link"
            // eslint-disable-next-line max-len
            href="#"
          >
            Learn more about Cloud Connector
          </Button>
          <Text component={TextVariants.h4}>Connection status of systems</Text>
        </TextContent>
        {isLoading ? (
          <Skeleton size="lg" />
        ) : (
          <Table
            variant={TableVariant.compact}
            aria-label="Systems"
            cells={[
              {
                title: 'Connection type',
                value: 'type',
              },
              {
                title: 'Systems',
                value: 'count',
              },
              isUserEntitled && {
                title: 'Connection status',
                value: 'status',
              },
            ]}
            rows={rows}
          >
            <TableHeader />
            <TableBody />
          </Table>
        )}
      </div>
    </Modal>
  );
};

ExecuteModal.propTypes = {
  isOpen: PropTypes.bool,
  onClose: PropTypes.func,
  showRefresh: PropTypes.bool,
  isLoading: PropTypes.bool,
  data: PropTypes.array,
  remediationId: PropTypes.string,
  issueCount: PropTypes.number,
  runRemediation: PropTypes.func,
  etag: PropTypes.string,
  setEtag: PropTypes.func,
  getEndpoint: PropTypes.func,
  sources: PropTypes.object,
};
