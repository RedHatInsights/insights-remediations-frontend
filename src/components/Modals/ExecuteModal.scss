@import '@patternfly/patternfly/sass-utilities/_all';

.ins-c-remediations-connection-status { margin-right: var(--pf-global--spacer--xs); }
.ins-c-remediations-connection-status-error { color: var(--pf-chart-global--danger--Color--100); }

@media only screen and (max-width: $pf-global--breakpoint--sm) {
    .pf-c-modal-box__footer {
        flex-wrap: wrap;
        .pf-c-button {
            margin-bottom: var(--pf-c-modal-box__footer--c-button--MarginRight);
        }
    }
}
