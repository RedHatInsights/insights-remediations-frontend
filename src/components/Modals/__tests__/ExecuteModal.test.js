/* eslint-disable camelcase */
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { ExecuteModal } from '../ExecuteModal';
import { Modal, Alert } from '@patternfly/react-core';

const data = [
  {
    executor_id: '722ec903-f4b5-4b1f-9c2f-23fc7b0ba390',
    executor_type: 'satellite',
    executor_name: 'Satellite 1 (connected)',
    system_count: 1,
    connection_status: 'connected',
  },
  {
    executor_id: null,
    executor_type: null,
    executor_name: null,
    system_count: 4,
    connection_status: 'no_executor',
  },
];

describe('Execute modal', () => {
  it('renders ExecuteModal component when given data', () => {
    const closeFn = jest.fn();

    const container = shallow(
      <ExecuteModal
        isOpen={true}
        onClose={closeFn}
        showRefresh={false}
        remediationId={'id'}
        data={data}
        etag={'etag'}
        isLoading={false}
        issueCount={1}
        remediationStatus={'initial'}
        runRemediation={() => null}
        setEtag={() => null}
      />
    );

    expect(container.find(Modal)).toHaveLength(1);
    expect(toJson(container)).toMatchSnapshot();
  });

  it('renders ExecuteModal with refresh message when showRefresh is true', () => {
    const closeFn = jest.fn();

    const container = shallow(
      <ExecuteModal
        isOpen={true}
        onClose={closeFn}
        showRefresh={true}
        remediationId={'id'}
        data={data}
        etag={'etag'}
        isLoading={false}
        issueCount={1}
        remediationStatus={'initial'}
        runRemediation={() => null}
        setEtag={() => null}
      />
    );

    expect(container.find(Modal)).toHaveLength(1);
    expect(container.find(Alert)).toHaveLength(1);
    expect(toJson(container)).toMatchSnapshot();
  });
});
