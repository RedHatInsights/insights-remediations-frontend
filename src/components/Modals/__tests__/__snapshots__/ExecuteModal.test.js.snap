// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Execute modal renders ExecuteModal component when given data 1`] = `
<Modal
  actions={
    Array [
      <Button
        isDisabled={true}
        onClick={[Function]}
        variant="primary"
      >
        Execute playbook on 0 system
      </Button>,
      <Button
        onClick={[Function]}
        variant="secondary"
      >
        Download playbook
      </Button>,
      null,
    ]
  }
  appendTo={[Function]}
  aria-describedby=""
  aria-label=""
  aria-labelledby=""
  className="remediations ins-c-execute-modal"
  hasNoBodyWrapper={false}
  isFooterLeftAligned={true}
  isOpen={true}
  onClose={[MockFunction]}
  ouiaSafe={true}
  showClose={true}
  title="Execute playbook"
  titleIconVariant={null}
  titleLabel=""
  variant="small"
>
  <div>
    <TextContent>
      <Text
        component="p"
      >
        Playbook contains 
        <b>
          1 action
        </b>
         affecting
        <b>
           
          5 systems.
           
        </b>
      </Text>
      <Text>
        <ExpandableSection
          className=""
          isActive={false}
          onToggle={[Function]}
          toggleText="About remote execution with Cloud connector"
          toggleTextCollapsed=""
          toggleTextExpanded=""
        >
          Playbooks can be executed on systems which:
          <List>
            <ListItem>
              Are connected to Insights via a Satellite instance which has Receptor/Cloud Connector enabled, or 
              <br />
              <Button
                className="pf-u-p-0"
                component="a"
                href="https://access.redhat.com/documentation/en-us/red_hat_insights/2020-04/html/remediating_issues_across_your_red_hat_satellite_infrastructure_using_red_hat_insights/configuring-your-satellite-infrastructure-to-communicate-with-insights"
                key="download"
                variant="link"
              >
                How to configure Receptor/Cloud Connector on Red Hat Satellite
              </Button>
            </ListItem>
            <ListItem>
              Are directly connected to Insights via Red Hat connector, and Cloud Connector is enabled 
              <br />
              <Button
                className="pf-u-p-0"
                href="#"
                key="configure"
                variant="link"
              >
                How to enable Cloud Connector with Red Hat connect
              </Button>
            </ListItem>
          </List>
        </ExpandableSection>
      </Text>
      <Text
        component="p"
      >
        Executed Ansible Playbooks run on eligible systems with Cloud Connector. The playbook will be pushed immediately after selecting “Execute playbook”. If the playbook has “Auto reboot” on, systems requiring reboot to complete an action will reboot.
      </Text>
      <Button
        className="pf-u-p-0"
        href="#"
        key="configure"
        variant="link"
      >
        Learn more about Cloud Connector
      </Button>
      <Text
        component="h4"
      >
        Connection status of systems
      </Text>
    </TextContent>
    <Table
      aria-label="Systems"
      borders={true}
      canSelectAll={true}
      canSortFavorites={true}
      cells={
        Array [
          Object {
            "title": "Connection type",
            "value": "type",
          },
          Object {
            "title": "Systems",
            "value": "count",
          },
          false,
        ]
      }
      className=""
      contentId="expanded-content"
      dropdownDirection="down"
      dropdownPosition="right"
      expandId="expandable-toggle"
      gridBreakPoint="grid-md"
      isStickyHeader={false}
      ouiaSafe={true}
      role="grid"
      rowLabeledBy="simple-node"
      rows={Array []}
      selectVariant="checkbox"
      variant="compact"
    >
      <TableHeader />
      <Component />
    </Table>
  </div>
</Modal>
`;

exports[`Execute modal renders ExecuteModal with refresh message when showRefresh is true 1`] = `
<Modal
  actions={
    Array [
      <Button
        isDisabled={true}
        onClick={[Function]}
        variant="primary"
      >
        Execute playbook on 0 system
      </Button>,
      <Button
        onClick={[Function]}
        variant="secondary"
      >
        Download playbook
      </Button>,
      null,
    ]
  }
  appendTo={[Function]}
  aria-describedby=""
  aria-label=""
  aria-labelledby=""
  className="remediations ins-c-execute-modal"
  hasNoBodyWrapper={false}
  isFooterLeftAligned={true}
  isOpen={true}
  onClose={[MockFunction]}
  ouiaSafe={true}
  showClose={true}
  title="Execute playbook"
  titleIconVariant={null}
  titleLabel=""
  variant="small"
>
  <div>
    <Alert
      isInline={true}
      title="The connection status of systems associated with this Playbook has changed. Please review again."
      variant="warning"
    />
    <TextContent>
      <Text
        component="p"
      >
        Playbook contains 
        <b>
          1 action
        </b>
         affecting
        <b>
           
          5 systems.
           
        </b>
      </Text>
      <Text>
        <ExpandableSection
          className=""
          isActive={false}
          onToggle={[Function]}
          toggleText="About remote execution with Cloud connector"
          toggleTextCollapsed=""
          toggleTextExpanded=""
        >
          Playbooks can be executed on systems which:
          <List>
            <ListItem>
              Are connected to Insights via a Satellite instance which has Receptor/Cloud Connector enabled, or 
              <br />
              <Button
                className="pf-u-p-0"
                component="a"
                href="https://access.redhat.com/documentation/en-us/red_hat_insights/2020-04/html/remediating_issues_across_your_red_hat_satellite_infrastructure_using_red_hat_insights/configuring-your-satellite-infrastructure-to-communicate-with-insights"
                key="download"
                variant="link"
              >
                How to configure Receptor/Cloud Connector on Red Hat Satellite
              </Button>
            </ListItem>
            <ListItem>
              Are directly connected to Insights via Red Hat connector, and Cloud Connector is enabled 
              <br />
              <Button
                className="pf-u-p-0"
                href="#"
                key="configure"
                variant="link"
              >
                How to enable Cloud Connector with Red Hat connect
              </Button>
            </ListItem>
          </List>
        </ExpandableSection>
      </Text>
      <Text
        component="p"
      >
        Executed Ansible Playbooks run on eligible systems with Cloud Connector. The playbook will be pushed immediately after selecting “Execute playbook”. If the playbook has “Auto reboot” on, systems requiring reboot to complete an action will reboot.
      </Text>
      <Button
        className="pf-u-p-0"
        href="#"
        key="configure"
        variant="link"
      >
        Learn more about Cloud Connector
      </Button>
      <Text
        component="h4"
      >
        Connection status of systems
      </Text>
    </TextContent>
    <Table
      aria-label="Systems"
      borders={true}
      canSelectAll={true}
      canSortFavorites={true}
      cells={
        Array [
          Object {
            "title": "Connection type",
            "value": "type",
          },
          Object {
            "title": "Systems",
            "value": "count",
          },
          false,
        ]
      }
      className=""
      contentId="expanded-content"
      dropdownDirection="down"
      dropdownPosition="right"
      expandId="expandable-toggle"
      gridBreakPoint="grid-md"
      isStickyHeader={false}
      ouiaSafe={true}
      role="grid"
      rowLabeledBy="simple-node"
      rows={Array []}
      selectVariant="checkbox"
      variant="compact"
    >
      <TableHeader />
      <Component />
    </Table>
  </div>
</Modal>
`;
